### PR TITLE
TG-4820 | Added support for http retries

### PR DESCRIPTION
**Pull request by liorr**
- JIRA Ticket URL: https://twingate.atlassian.net/browse/TG-4820

## Changes
- Added support for HTTP retries
- Provider configuration for HTTP timeout and max retry

### CODE REVIEW

File: twingate/client.go [hunk 1]
```diff
@@ -60,24 +56,24 @@ func NewAPIError(wrappedError error, operation string, resource string) *APIErro
 }
 
 func (e *APIError) Error() string {
-	var a = make([]interface{}, 0, 2) //nolint:gomnd
-	a = append(a, e.Operation, e.Resource)
+	var args = make([]interface{}, 0, 2) //nolint:gomnd
```
Comment (liorr) [inline twingate/client.go:59]: those are new lint errors fixes

File: twingate/client.go [hunk 2]
```diff
@@ -105,23 +102,23 @@ type Client struct {
...
-	Version             string
+	underlineRoundTripper http.RoundTripper
+	APIToken              string
+	Version               string
 }
 
 func (t *transport) RoundTrip(req *http.Request) (*http.Response, error) {
-	req.Header.Add("X-API-KEY", t.APIToken)
-	req.Header.Add("User-Agent", fmt.Sprintf("TwingateTF/%s", t.Version))
+	req.Header.Set("X-API-KEY", t.APIToken)
```
Comment (liorr) [inline twingate/client.go:111]: this is called again on retry and add will append an additional value on every retry

File: twingate/client.go [hunk 3]
```diff
@@ -144,22 +141,26 @@ func newServerURL(network, url string) serverURL {
...
 	client := Client{
-		HTTPClient:       retryableClient,
+		RetryableClient:  retryableClient,
+		HTTPClient:       httpClient,
 		ServerURL:        sURL.url,
 		GraphqlServerURL: sURL.newGraphqlServerURL(),
 		APIServerURL:     sURL.newAPIServerURL(),
 		APIToken:         apiToken,
-		GraphqlClient:    graphql.NewClient(sURL.newGraphqlServerURL(), retryableClient.HTTPClient),
+		GraphqlClient:    graphql.NewClient(sURL.newGraphqlServerURL(), httpClient),
```
Comment (liorr) [inline twingate/client.go:163]: this is the fix, to get a standard HTTP client that supports retries, you need to get an instance from  [retryableClient.StandardClient()](https://github.com/Twingate/terraform-provider-twingate/pull/51/files#diff-134984f9bebbd0ae1f657dc170e63d3c218ac4d72f03ac098fabeb53c7a87183R154)

File: twingate/client_connector_test.go [hunk 4]
```diff
@@ -171,54 +170,6 @@ func TestClientConnectorUpdateErrorWhenIdEmpty(t *testing.T) {
 	})
 }
 
-func TestClientConnectorUpdateErrorWhenMutationWrong(t *testing.T) {
```
Comment (liorr) [inline twingate/client_connector_test.go:174]: I moved some tests around to organize the file better

File: twingate/client_connector_tokens.go [hunk 5]
```diff
@@ -17,13 +17,20 @@ type connectorTokens struct {
 
 const connectorTokensResourceName = "connector tokens"
 
-func (client *Client) verifyConnectorTokens(refreshToken, accessToken string) error {
-	jsonValue, _ := json.Marshal(
+func (client *Client) verifyConnectorTokens(ctx context.Context, refreshToken, accessToken string) error {
```
Comment (liorr) [inline twingate/client_connector_tokens.go:20]: a few new lint fixes here (error handling and support for context)

File: twingate/twingate_sweeper_test.go [hunk 6]
```diff
@@ -26,7 +27,13 @@ func sharedClient(tenant string) (*Client, error) {
 		return nil, fmt.Errorf("must provide environment variable TWINGATE_URL")
 	}
 
-	client := NewClient(os.Getenv("TWINGATE_URL"), os.Getenv("TWINGATE_API_TOKEN"), os.Getenv("TWINGATE_NETWORK"), "sweeper")
+	client := NewClient(
+		os.Getenv("TWINGATE_URL"),
+		os.Getenv("TWINGATE_API_TOKEN"),
+		os.Getenv("TWINGATE_NETWORK"),
+		30*time.Second,
```
Comment (romankor) [inline twingate/twingate_sweeper_test.go:34]: Lets use 
```golang
value = os.getenv("TWINGATE_HTTP_TIMEOUT", 30*time.Second)
```

Give the option to override that from outside the test 
same for the second variable

File: twingate/client_connector_test.go [hunk 7]
```diff
@@ -464,3 +363,65 @@ func TestClientConnectorReadAllOk(t *testing.T) {
...
+
+		client := newHTTPMockClient()
+		defer httpmock.DeactivateAndReset()
+		httpmock.RegisterResponder("POST", client.GraphqlServerURL,
+			httpmock.NewErrorResponder(errors.New("error_1")))
+		connectorId := graphql.ID("test")
+
+		err := client.updateConnector(context.Background(), connectorId, "new name")
+
+		assert.EqualError(t, err, "failed to update connector with id test: Post \"https://test.twindev.com/api/graphql/\": error_1")
```
Comment (romankor) [inline twingate/client_connector_test.go:378]: Can't figure out why is expecting this specific URL `test.twindev.com` and where it is configured

**Review by romankor — APPROVED** [review]
Had a comment on the test provider initialization , and a question about where the test URL is coming from , other than that didnt find anything else 

I still think its better to have linter pinned to a specific version , that way we dont have to deal with these changes in different PR's
linter will be updated by dependabot and all the updates to the code will be done in the dependabot PR 
So Fix and Merge

File: twingate/twingate_sweeper_test.go (hunk 6 above)
Comment (liorr) [reply twingate/twingate_sweeper_test.go:34]: There is no builtin support for default value when getting env vars but I have added it

File: twingate/client_connector_test.go (hunk 7 above)
Comment (liorr) [reply twingate/client_connector_test.go:378]: We set up everything at ` newHTTPMockClient` - https://github.com/Twingate/terraform-provider-twingate/blob/37fd097212777e9e23da5db472464086f2a1b87c/twingate/client_test.go#L10

**Pull request merged**